### PR TITLE
feat: Added Facebook OAuth

### PR DESCRIPTION
**Pull request by mmubarak0**
## The code has been verified as safe.🛩 ✅
Implemented an endpoint for user login via social media accounts. It manages all server-to-server communications required to authenticate the user.

> This code implements social authentication safely, ensuring no harm to the database model. It creates a separate entity named `OAuthUser`, which is associated with the user model through a relationship, leaving the `User` model unmodified.

## Related Issue:

- https://github.com/hngprojects/hng_boilerplate_python_fastapi_web/issues/55
​

## Description
The endpoint `/api/v1/auth/facebook-login` is created for user authentication via their Facebook account. It ensures a secure login by interfacing with the Facebook API, providing a seamless user experience.

## Testing

The endpoint has undergone testing via Postman and PyTest, The resulting images are provided below at the end.

#### To run test cases use the following command

```py
python -m pytest tests/v1/test_facebook_auth.py
```

## Steps to test the endpoint

- first go to https://developers.facebook.com/tools/explorer
- Login with your developer account if asked.
- generate a new access token an pass
- use your newly created access token to test the endpoint 
- here is a curl command if prefer to use command line
```bash
curl -X POST -H "Content-Type: application/json" -d '{"access_token": "YOUR_ACCESS_TOKEN"}' http://example.com/api/v1/auth/facebook-login
```

## Types of changes

- [x] Integrate Facebook OAuth API.
- [x] Create backend endpoints to handle OAuth authentication ``` api/v1/auth/facebook-login ```
- [x] Secure storage of OAuth tokens.
- [x] Ensure payload/response objects as follow ``` { "message": string, "status_code": number } ```
- [x] Ensure compliance with Facebook’s OAuth policies.
- [x] All new and existing test are passed.

#### successful login

> 201 response: the user is login in for the first time
image
> 200 response: the is already registered on the database
image

image

#### missing token

> 422 response: the token is missing from the response but the body is passed correctly
image
> 422 response: the body has not been passed correctly
image

### CODE REVIEW

File: tests/v1/user_deactivation_test.py [hunk 1]
```diff
@@ -1,8 +1,8 @@
-import sys, os
-import warnings
+# import sys, os
+# import warnings
 
-warnings.filterwarnings("ignore", category=DeprecationWarning)
-sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '../..')))
+# warnings.filterwarnings("ignore", category=DeprecationWarning)
+# sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '../..')))
```
Comment (Goketech) [inline tests/v1/user_deactivation_test.py:5]: revert this

File: tests/v1/user_deactivation_test.py [hunk 2]
```diff
@@ -23,9 +23,9 @@
 
 @pytest.fixture
 def mock_db_session():
-    """Fixture to create a mock database session."""
+    """Fixture to create a mock database session. api.v1.services.user.get_db"""
 
-    with patch("api.v1.services.user.get_db", autospec=True) as mock_get_db:
+    with patch("api.db.database.get_db", autospec=True) as mock_get_db:
         mock_db = MagicMock()
```
Comment (Goketech) [inline tests/v1/user_deactivation_test.py:29]: why did you update this file?

File: api/utils/settings.py [hunk 3]
```diff
@@ -31,5 +31,7 @@ class Settings(BaseSettings):
     MAIL_PORT: int = config('MAIL_PORT')
     MAIL_SERVER: str = config('MAIL_SERVER')
 
+    # FACEBOOK_APP_ID: str = config("FACEBOOK_APP_ID")
+    # FACEBOOK_APP_SECRET: str = config("FACEBOOK_APP_SECRET")
```
Comment (Goketech) [inline api/utils/settings.py:35]: remove your comments

File: api/v1/routes/blog.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+from api.db.database import get_db
+
+blog = APIRouter(prefix="/blogs", tags=["Blog"])
+
+@blog.get("/", response_model=List[BlogResponse])
+def get_all_blogs(db: Session = Depends(get_db)):
+    blogs = db.query(Blog).filter(Blog.is_deleted == False).all()
+    if not blogs:
+        return []
+    return blogs
```
Comment (Goketech) [inline api/v1/routes/blog.py:16]: title of pr is "added facebook OAuth"

why making changes to blog.py

File: api/v1/models/profile.py [hunk 5]
```diff
@@ -16,9 +16,6 @@ class Profile(BaseTableModel):
     __tablename__ = 'profiles'
 
     user_id = Column(String, ForeignKey('users.id', ondelete="CASCADE"), unique=True, nullable=False)
-    username = Column(String, nullable=True)
-    first_name = Column(String, nullable=True)
-    last_name = Column(String, nullable=True)
```
Comment (Goketech) [inline api/v1/models/profile.py:21]: why editing the model

**Review by Goketech — CHANGES_REQUESTED** [review]

File: tests/v1/user_deactivation_test.py (hunk 1 above)
Comment (mmubarak0) [reply tests/v1/user_deactivation_test.py:5]: I am not the one who made those changes

Comment (mmubarak0) [reply tests/v1/user_deactivation_test.py:5]: this beacause the `auth-v1` branch is not synced with `dev`

Comment (mmubarak0) [reply tests/v1/user_deactivation_test.py:5]: Do I have to change my PR to `dev` ?

File: api/v1/routes/blog.py (hunk 4 above)
Comment (mmubarak0) [reply api/v1/routes/blog.py:16]: This is not me again. bruh!

File: api/v1/models/profile.py (hunk 5 above)
Comment (mmubarak0) [reply api/v1/models/profile.py:21]: also not me ;)

File: tests/v1/user_deactivation_test.py (hunk 1 above)
Comment (Goketech) [reply tests/v1/user_deactivation_test.py:5]: it shows you made changes to the file
revert it on local
pull from dev branch and make sure your branch is updated

File: api/v1/routes/blog.py (hunk 4 above)
Comment (Goketech) [reply api/v1/routes/blog.py:16]: it shows you made changes to the file, revert changes and pull from dev branch

File: api/v1/models/profile.py (hunk 5 above)
Comment (SundayMba) [reply api/v1/models/profile.py]: @Goketech most of the changes wasnt his.

File: api/utils/settings.py (hunk 3 above)
Comment (mmubarak0) [reply api/utils/settings.py:35]: this is not a comment in order for my endpoint to work those should be uncommented. I will leave that for the devops guy

Comment (mmubarak0) [reply api/utils/settings.py:35]: Can I resolve this now?

**Comment by Goketech** [comment]
@joboy-dev check this

**Comment by mmubarak0** [comment]
@buka please at least a hint ? where is the error i.e description / naming / branch-name / etc .....

**Comment by mmubarak0** [comment]
@Goketech can you mark the requested changes as resolved ?

**Pull request closed without merging**